### ISSUE TITLE
Unique appends and inserts

### ISSUE THREAD

**Issue by ykonoclast** [xonsh/xonsh#2562]
It may be useful for Xonsh to provide for append and insert-like commands but with duplicate checking. 

For instance, such a unique append could be used to add directories to $PATH while guaranteeing there would be no duplicate entry (this can happen when using rc-files since Xonsh may have been launched by the environment prior to the user opening a terminal window).

**Comment by agarick** [comment]
Just switched to xonsh (wooh!) and came across the "appended twice" issue. The solution I've opted for checks uniqueness, not during, but after insertion, so I'm not sure if it will help.

```python
for i in ['/usr/local/bin', '/opt/bin', '/usr/sbin', '/sbin']: $PATH.insert(0, i)

from collections import OrderedDict
$PATH = list(OrderedDict.fromkeys($PATH))
```
ref: https://stackoverflow.com/questions/480214/how-do-you-remove-duplicates-from-a-list-in-whilst-preserving-order

**Comment by astronouth7303** [comment]
@agarick just use `set`

This was also discussed in #2537.

**Comment by adqm** [comment]
@astronouth7303: `set` won't preserve order, though, which is important for the path.  Using `OrderedDict` is a cute trick, I think.  But another alternative would be to do a check at insertion time:
```
for i in ['/usr/local/bin', '/opt/bin', '/usr/sbin', '/sbin']:
    if i not in $PATH:
        $PATH.insert(0, i)
```

@ykonoclast: I think this is an interesting idea.  I think there's a little bit of ambiguity, though.  Imagine that a "unique append" is performed with something that already exists in the path.  I could imagine one of two behaviors: either it moves that thing to the end of the path (so that the result is the same as if we had appended it to a path that didn't contain the element to begin with), _or_ it could simply leave the path unchanged.  Maybe it's inconsequential, but I'm not sure which of those feels more natural.

**Comment by bestlem** [comment]
It is ambigous as to what to do with an append but a prepend is less of a problem as the new one you are adding is in the correct place. I use

    def __path_prepend(path, new_paths):
	 new_paths.reverse()
	 for p in new_paths:
		p1 = Path(p).expanduser()
		if p1.as_posix() in path:
			path.remove(p1.as_posix())
		if p1.exists():
			path.insert(0, p)

I also check the path exists.  
So

    >>> $XPATH = '/usr/local/bin:/usr/bin:/bin'
     >>> __path_prepend($XPATH, ['~/bin', '/usr/local/bin', '/opt/local/bin'])
    >>> $XPATH                                             
    EnvPath(
    ['/Users/mark/bin',
     '/usr/local/bin',
     '/opt/local/bin',
     '/usr/bin',
     '/bin']
    )

**Comment by scopatz** [comment]
Still happy to have a PR that implements a unique append / preprend on the `xonsh.tools.EnvPath` class.

**Comment by gsaga** [comment]
@scopatz What kind of behavior is intended:

1. Don't add if already present

2. Move to start/end if present

**Comment by scopatz** [comment]
@gsaga - I can image people wanting both behaviors, and both being valid. It is probably best to have a kwarg that allows users to pick.  I think (1) should probably be the default, but I don't mind either way :)

**Comment by scopatz** [comment]
This has been implemented in #2664
